lightning/src/onion_message/functional_tests.rs

```diff
@@ -427,51 +462,97 @@ fn many_hops() {
 	};
-	nodes[0].messenger.send_onion_message(path, test_msg, None).unwrap();
+	nodes[0].messenger.send_onion_message_using_path(path, test_msg, None).unwrap();
 	nodes[num_nodes-1].custom_message_handler.expect_message(TestCustomMessage::Response);
 	pass_along_path(&nodes);
 }
 
 #[test]
-fn spec_test_vector() {
-	let keys_mgrs = vec![
-		(Arc::new(test_utils::TestKeysInterface::new(&[0; 32], Network::Testnet)), // Alice
-		 Arc::new(test_utils::TestNodeSigner::new(SecretKey::from_slice(&<Vec<u8>>::from_hex("4141414141414141414141414141414141414141414141414141414141414141").unwrap()).unwrap()))),
-		(Arc::new(test_utils::TestKeysInterface::new(&[1; 32], Network::Testnet)), // Bob
-		 Arc::new(test_utils::TestNodeSigner::new(SecretKey::from_slice(&<Vec<u8>>::from_hex("4242424242424242424242424242424242424242424242424242424242424242").unwrap()).unwrap()))),
-		(Arc::new(test_utils::TestKeysInterface::new(&[2; 32], Network::Testnet)), // Carol
-		 Arc::new(test_utils::TestNodeSigner::new(SecretKey::from_slice(&<Vec<u8>>::from_hex("4343434343434343434343434343434343434343434343434343434343434343").unwrap()).unwrap()))),
-		(Arc::new(test_utils::TestKeysInterface::new(&[3; 32], Network::Testnet)), // Dave
-		 Arc::new(test_utils::TestNodeSigner::new(SecretKey::from_slice(&<Vec<u8>>::from_hex("4444444444444444444444444444444444444444444444444444444444444444").unwrap()).unwrap()))),
-	];
-	let message_router = Arc::new(TestMessageRouter {});
-	let offers_message_handler = Arc::new(TestOffersMessageHandler {});
-	let custom_message_handler = Arc::new(TestCustomMessageHandler::new());
-	let mut nodes = Vec::new();
-	for (idx, (entropy_source, node_signer)) in keys_mgrs.iter().enumerate() {
-		let logger = Arc::new(test_utils::TestLogger::with_id(format!("node {}", idx)));
-		nodes.push(OnionMessenger::new(
-			entropy_source.clone(), node_signer.clone(), logger.clone(), message_router.clone(),
-			offers_message_handler.clone(), custom_message_handler.clone()
-		));
+fn requests_peer_connection_for_buffered_messages() {
+	let nodes = create_nodes(3);
+	let message = TestCustomMessage::Request;
+	let secp_ctx = Secp256k1::new();
+	let blinded_path = BlindedPath::new_for_message(
+		&[nodes[1].node_id, nodes[2].node_id], &*nodes[0].entropy_source, &secp_ctx
+	).unwrap();
+	let destination = Destination::BlindedPath(blinded_path);
+
+	// Buffer an onion message for a connected peer
+	nodes[0].messenger.send_onion_message(message.clone(), destination.clone(), None).unwrap();
+	assert!(release_events(&nodes[0]).is_empty());
+	assert!(nodes[0].messenger.next_onion_message_for_peer(nodes[1].node_id).is_some());
+	assert!(nodes[0].messenger.next_onion_message_for_peer(nodes[1].node_id).is_none());
+
+	// Buffer an onion message for a disconnected peer
+	disconnect_peers(&nodes[0], &nodes[1]);
+	assert!(nodes[0].messenger.next_onion_message_for_peer(nodes[1].node_id).is_none());
+	nodes[0].messenger.send_onion_message(message, destination, None).unwrap();
```

Comment:
This would causes us to hit the debug assertion in `dequeue_message`. We don't prevent someone from calling `next_onion_message_for_peer` for a disconnected peer. Perhaps we should return `None`? Calling it means the `PeerManager` state is out-of-sync with the `OnionMessenger` state.